waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/DataFlow.scala

```diff
@@ -359,4 +404,69 @@ case class DataFlowActionTags(tags: Set[String], dependentOnTags: Set[String])
-case class DataFlowTagState(activeTags: Set[String], activeDependentOnTags: Set[String], taggedActions: Map[String, DataFlowActionTags])
\ No newline at end of file
+case class DataFlowTagState(activeTags: Set[String], activeDependentOnTags: Set[String], taggedActions: Map[String, DataFlowActionTags])
+
+/** When a Data Flow is defined, certain hints related to its execution can be specified, these hints will help scheduler
+  * with deciding when and where to run the action. Further uses can be added to it.
+  *
+  * At the moment, when an action is added to the scheduling meta, it will automatically assign it the current Execution
+  * Pool, but if there were other global context attributes to assign, than the action could aquire them as well.
+  *
+  * @param state           describes a current state of schedulingMeta
+  * @param actionState     Map[DataFlowAction.schedulingGuid, Execution Pool Name] - association between actions and execution pool names
+  * @tparam C
+  */
+case class SchedulingMeta[C](state: SchedulingMetaState, actionState: Map[String, SchedulingMetaState]) {
+
+  def this() = this(SchedulingMetaState(DEFAULT_POOL_NAME), Map.empty)
+
+  /**
+    * Adds action to the scheduling meta, action aquires all of the relevant context attributes (like currentExecutionPoolName)
+    *
+    * @param action action to add to the scheduling meta
+    * @return       new state of the scheduling meta with action associated with relevant context attributes
+    */
+  def addAction(action: DataFlowAction[C]): SchedulingMeta[C] = {
+    SchedulingMeta(state, actionState + (action.schedulingGuid -> state))
+  }
+
+  /**
+    * Removes the action from scheduling meta.
+    *
+    * @param action
+    * @return       new state of the scheduling meta without the action
+    */
+  def removeAction(action: DataFlowAction[C]): SchedulingMeta[C] = {
+    SchedulingMeta(state, actionState - action.schedulingGuid)
+  }
+
+  /**
+    * Gets action's execution pool name.
+    *
+    * @param action
+    * @return       execution pool name of the action, if not found than returns DEFAULT_POOL_NAME
+    */
+  def executionPoolName(action: DataFlowAction[C]): String = actionState.get(action.schedulingGuid).map(_.executionPoolName).getOrElse(DEFAULT_POOL_NAME)
+
+  /**
+    * Sets current pool name into the context of the scheduling meta.
+    *
+    * @param newState
+    * @return                   new state of the scheduling meta with new execution pool name, all subsequent actions will be added to it.
+    */
+  def setState(newState: SchedulingMetaState): SchedulingMeta[C] = SchedulingMeta(newState, actionState)
+
+}
+
+/**
+  * Contains values that will be associated with all actions added to the data flow.
+  *
+  * @param executionPoolName  name of the execution pool
+  */
+case class SchedulingMetaState(executionPoolName: String) {
+
+  //TODO: May be add tags into here to have a common place to accumulate extra scheduling info about actions and
```

Comment:
Feel like you got halway through this TODO comment?

Reply:
yep, added word tags